test_palindrome.py

```diff
@@ -0,0 +1,151 @@
+import pytest
+from collections import namedtuple
+from palindrome import is_palindrome
+
+
+@pytest.mark.parametrize("case_sensitive", [True, False])
+@pytest.mark.parametrize("whitespace_sensitive", [True, False])
+@pytest.mark.parametrize("punctuation_sensitive", [True, False])
+def test_non_string_data_raises_valueerror(
+        case_sensitive: bool,
+        whitespace_sensitive: bool,
+        punctuation_sensitive: bool,
+) -> None:
+    TestCaseNoOptions = namedtuple(
+        typename='TestCaseNoOptions',
+        field_names=["expectation", "data"]
+    )
+    test_cases = [
+        TestCaseNoOptions(expectation="Expected string, got <class 'NoneType'>!", data=None),
+        TestCaseNoOptions(expectation="Expected string, got <class 'bytes'>!", data=b""),
+        TestCaseNoOptions(expectation="Expected string, got <class 'bytearray'>!", data=bytearray()),
+        TestCaseNoOptions(expectation="Expected string, got <class 'tuple'>!", data=tuple()),
+        TestCaseNoOptions(expectation="Expected string, got <class 'list'>!", data=[]),
+        TestCaseNoOptions(expectation="Expected string, got <class 'int'>!", data=0),
+        TestCaseNoOptions(expectation="Expected string, got <class 'float'>!", data=0.0),
+    ]
+    for expectation, data in test_cases:
+        with pytest.raises(ValueError) as value_error:
+            is_palindrome(data, case_sensitive, whitespace_sensitive, punctuation_sensitive)
+        assert str(value_error.value) == expectation, \
+            f"Expected '{expectation}', got '{str(value_error.value)}' for '{data=}' and "\
+            f"options '{case_sensitive=}', '{whitespace_sensitive=}', '{punctuation_sensitive=}'!"
+
+
+@pytest.mark.parametrize("case_sensitive", [True, False])
+@pytest.mark.parametrize("whitespace_sensitive", [True, False])
+@pytest.mark.parametrize("punctuation_sensitive", [True, False])
+def test_empty_string_is_a_palindrome(
+        case_sensitive: bool,
+        whitespace_sensitive: bool,
+        punctuation_sensitive: bool,
+) -> None:
+    expectation, data = True, ""
+    result = is_palindrome(data, case_sensitive, whitespace_sensitive, punctuation_sensitive)
+    assert result == expectation,\
+        f"Expected '{expectation}', got '{result}' for '{data=}' and "\
+        f"options '{case_sensitive=}', '{whitespace_sensitive=}', '{punctuation_sensitive=}'!"
+
+
+@pytest.mark.parametrize("case_sensitive", [True, False])
+@pytest.mark.parametrize("whitespace_sensitive", [True, False])
+@pytest.mark.parametrize("punctuation_sensitive", [True, False])
+def test_strict_palindromes_that_are_not_affected_by_options(
+        case_sensitive: bool,
+        whitespace_sensitive: bool,
+        punctuation_sensitive: bool,
+) -> None:
+    TestCaseNoOptions = namedtuple(
```

Comment:
Niby można tworzyć takie nazwane krotki w testach, chociaż ja osobiście skłaniam się ku jaknajbardziej prymitywnym testom - tak żeby od razu było wiadomo co pada, bez konieczności zapoznawania się z tym jak test jest napisany.

P.S. Polecam także taką opcję deklarowania nazwanych krotek jak tu:
https://stackoverflow.com/questions/50766461/whats-the-difference-between-namedtuple-and-namedtuple